src/backend/optimizer/path/allpaths.c

```diff
@@ -2110,15 +2110,43 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 			subplan = subquery_planner(cteroot->glob, subquery, cteroot, cte->cterecursive,
 									   tuple_fraction, &subroot, config);
 
-			cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
+			/*
+			 * Sharing General and SegmentGeneral subplan may lead to deadlock
+			 * when executed with 1-gang and joined with n-gang.
+			 */
+			if (CdbPathLocus_IsGeneral(*subplan->flow) ||
+			    CdbPathLocus_IsSegmentGeneral(*subplan->flow))
+			{
+				cteplaninfo->subplan = subplan;
+			}
+			else
+			{
+				cteplaninfo->subplan = prepare_plan_for_sharing(cteroot, subplan);
+			}
+
 			cteplaninfo->subroot = subroot;
 		}
 
 		/*
 		 * Create another ShareInputScan to reference the already-created
-		 * subplan.
+		 * subplan if not avoiding sharing for General and SegmentGeneral
```

Comment:
May be move comment inside if-block?

Reply:
It has information about execution in case of not entering the if block